src/rust/engine/process_execution/src/local.rs

```diff
@@ -234,12 +237,36 @@ impl super::CommandRunner for CommandRunner {
       .store
       .materialize_directory(workdir_path.clone(), req.input_files)
       .and_then(move |()| {
-        if let Some(jdk_home) = maybe_jdk_home {
-          symlink(jdk_home, workdir_path3.join(".jdk"))
+        let jdk_home_symlinked: Result<(), _> = maybe_jdk_home.map_or(Ok(()), |jdk_home| {
+          symlink(jdk_home, workdir_path3.clone().join(".jdk"))
             .map_err(|err| format!("Error making symlink for local execution: {:?}", err))
-        } else {
-          Ok(())
-        }
+        });
+        // The bazel remote execution API specifies that the parent directories for output files and
+        // output directories should be created before execution completes: see
+        // https://github.com/pantsbuild/pants/issues/7084.
+        let parent_paths_created: Result<(), _> = output_file_paths2
+          .iter()
+          .chain(output_dir_paths2.iter())
+          // Path#parent() returns None if the parent is root on unix.
+          .filter_map(|rel_path| rel_path.parent())
+          .map(|parent_relpath| workdir_path3.join(parent_relpath))
+          // TODO: we use a HashSet to deduplicate directory paths to create, but it would probably
+          // be even more efficient to only retain the directories at greatest nesting depth, as
+          // create_dir_all() will ensure all parents are created. At that point, we might consider
+          // explicitly enumerating all the directories to be created and just using create_dir(),
+          // unless there is some optimization in create_dir_all() that makes that less efficient.
+          .collect::<HashSet<_>>()
```

Comment:
I think this would be more clear as:
```rust
let parent_paths_to_create: HashSet<_> = stuff.collect();
for path in parent_paths_to_create {
  create_dir_all(&path).map_err(...)?;
}
Ok(())
```

rather than using a chain of two iter-collect pairs.

Reply:
github lost my reply somehow! I wanted to say thank you because as much as I love `.collect()` this is significantly easier to write and read and I wasn't really remembering the utility of `?`!